### ISSUE TITLE
Incorrect suspicious_op_assign_impl

### ISSUE THREAD

**Issue by kjeremy** [rust-lang/rust-clippy#5255]
The following triggers this lint and probably should not
https://github.com/rust-analyzer/rust-analyzer/blob/39a9e42bf8d363e83d92bedf0ad07e95a42b1602/crates/rust-analyzer/src/semantic_tokens.rs#L68

clippy 0.0.212 (69f99e74 2019-12-14)

**Comment by kjeremy** [comment]
See https://github.com/rust-analyzer/rust-analyzer/pull/3409#discussion_r386618929

**Comment by jpospychala** [comment]
The issue here is that AssignOp kind is ignored, but expression's top level operator is checked when triggering the lint.

Basic solution is to ignore whole all operators in expression if it's is part of AssignOp.

But it's also noticeable that AssignOp kind is not looked at while it should so lint doesn't detect following mistake:
```rust
impl AddAssign for Foo {
    fn add_assign(&mut self, other: Foo) {
        *self -= other;
    }
}
```

but it detects this one:

```rust
impl AddAssign for Foo {
    fn add_assign(&mut self, other: Foo) {
        *self = *self - other ;
    }
}
```
